src/app-layer-ftp.c

```diff
@@ -378,6 +470,30 @@ static void FtpTransferCmdFree(void *data)
     FTPFree(cmd, sizeof(struct FtpTransferCmd));
 }
 
+static uint32_t CopyCommandLine(uint8_t **dest, uint8_t *src, uint32_t length)
+{
+    if (likely(length)) {
+        if (unlikely(length == UINT32_MAX)) {
```

Comment:
what is this for? Does length == UINT32_MAX have a special meaning somehow?

Reply:
Integer overflow check for allocation of `length+1` bytes and copy of `length` bytes, in order to have `where` always be null terminated.  Other constraints may remove the need for that check on line 476, i.e., if `length` cannot approach UINT32_MAX.  Or perhaps `where` should remain conditionally null terminated as it had been, subject to whitespace being found.  @jlucovsky, guidance?

Reply:
`length` will never approach `UINT32_MAX` -- it's a subset of the actual input line received, further split into individual FTP request commands.

Reply:
Earlier it looked to me like `FTPParseResponse` (one of the callers of `CopyCommandLine`) was passing in raw reassembled stream.  Either way, is it okay - low overhead - to leave the overflow check in?  And with the talk of `strlen` elsewhere in this issue, it may be good to ensure at time of copy that the `tx->response_list` elements (allocated as `where`) are null terminated even in absence of trailing whitespace, which drove the `length+1` (vs. previous `length`) allocation.  Again, okay to leave in?

(And as always, @jlucovsky is also encouraged to make further edits Jeff deems prudent.  Also happy to discuss further, and to make changes either of you request.)

Reply:
I don't object to some kind of overflow guarding, but this seems very limited. What about checking that allocations by the FTP parser can never be more than 10MiB/100MiB/1GiB (which ever seems prudent)? This could then move into the FTPMalloc/FTPCalloc/FTPRealloc wrappers.

Reply:
That check is indeed limited, but unless `length` is otherwise capped itself, that overflow check is still reasonable due to allocating and writing a different number of bytes here.

e.g., if there did happen to be an internal `FTPCalloc` cap, `UINT32_MAX+1` will be less than it, and POSIX permits `calloc` (and `malloc`) to return a non-NULL pointer for zero size allocations.

Reply:
I'd think a follow up issue is warranted to provide size checking in the alloc routines throughout.